In [42]:
import pandas as pd
df = pd.read_excel("week5_kigali_weather.xlsx")
df.head()

,day,date,temp_c,rainfall_mm,humidity_pct
0,1,46266,24.4,0.0,77.0
1,2,46267,25.9,0.0,74.7
2,3,46268,24.5,0.0,72.1
3,4,46269,26.0,0.0,76.3
4,5,46270,24.6,22.9,73.2


In [43]:
temps = [24.4, 25.9, 28.6, 19.3, 26.0, 22.1]
raw = ["24.4", "25.9", "28.6", "19.3"]

In [44]:
temps_f = [round(t * 9/5 + 32, 2) for t in temps]
print(temps_f)

[75.92, 78.62, 83.48, 66.74, 78.8, 71.78]


In [45]:
# (a) temperatures >= 26
hot = [t for t in temps if t >= 26]
print(hot)
# [28.6, 26.0]

# (b) temperatures below 24, rounded
cool_rounded = [round(t) for t in temps if t < 24]
print(cool_rounded)
# [19, 22]

[28.6, 26.0]
[19, 22]


In [46]:
floats = [float(s) for s in raw]
print(floats)
# [24.4, 25.9, 28.6, 19.3]

print(sum(floats) / len(floats))
# 24.55

[24.4, 25.9, 28.6, 19.3]
24.55


In [47]:
results = [("Aline", 85), ("Eric", 78), ("Diane", 91), ("Samuel", 92), ("Grace", 76)]

ranked = sorted(results, key=lambda r: r[1], reverse=True)
print(ranked[:3])

[('Samuel', 92), ('Diane', 91), ('Aline', 85)]


In [48]:
rain_log = [(3, 0.0), (7, 12.5), (11, 3.2), (14, 26.2), (18, 8.9)]

wettest = max(rain_log, key=lambda d: d[1])
print(wettest)

(14, 26.2)


In [49]:
df = pd.read_excel("week5_kigali_weather.xlsx")
temps = df["temp_c"].to_numpy()
dates = df["date"].to_numpy()

In [50]:
import numpy as np
import time

n = 1_000_000
py_list = list(range(n))
arr = np.arange(n)

start = time.perf_counter()
doubled_list = [x * 2 for x in py_list]
t_python = time.perf_counter() - start

start = time.perf_counter()
doubled_arr = arr * 2
t_numpy = time.perf_counter() - start

print(f"Python loop: {t_python:.4f} s")
print(f"NumPy:       {t_numpy:.4f} s")
print(f"NumPy was about {t_python / t_numpy:.0f}x faster")

Python loop: 0.3077 s
NumPy:       0.0127 s
NumPy was about 24x faster


In [51]:
print(f"Mean: {temps.mean():.2f} °C")
print(f"Max:  {temps.max():.2f} °C")
print(f"Min:  {temps.min():.2f} °C")
print(f"Std:  {temps.std():.2f} °C")

Mean: 24.00 °C
Max:  28.60 °C
Min:  19.30 °C
Std:  2.61 °C


In [52]:
pos = np.argmax(temps)
print(f"Hottest day: {dates[pos]} at {temps[pos]} °C")

Hottest day: 46282 at 28.6 °C


In [53]:
df = pd.read_excel("week5_kigali_weather.xlsx")
temps = df["temp_c"].to_numpy()
rain = df["rainfall_mm"].to_numpy()
humidity = df["humidity_pct"].to_numpy()

In [54]:
hot_mask = temps >= 26

print(f"(a) Hot days: {hot_mask.sum()}")
print(f"(b) Average temp on hot days: {temps[hot_mask].mean():.2f} °C")

(a) Hot days: 18
(b) Average temp on hot days: 27.11 °C


In [55]:
rainy_mask = rain > 0

print(f"(a) Rainy days: {rainy_mask.sum()}")
print(f"(b) Total rainfall: {rain.sum():.1f} mm")
print(f"(c) Heaviest single day: {rain.max():.1f} mm")

(a) Rainy days: 16
(b) Total rainfall: 331.2 mm
(c) Heaviest single day: 26.2 mm


In [56]:
# (a) rainy vs dry average temperature
print(f"(a) Rainy days avg temp: {temps[rain > 0].mean():.2f} °C")
print(f"    Dry days avg temp:   {temps[rain == 0].mean():.2f} °C")

# (b) hot AND rainy
print(f"(b) Hot and rainy days: {((temps >= 26) & (rain > 0)).sum()}")

# (c) humid days
print(f"(c) Days with humidity >= 70: {(humidity >= 70).sum()}")

(a) Rainy days avg temp: 24.14 °C
    Dry days avg temp:   23.95 °C
(b) Hot and rainy days: 5
(c) Days with humidity >= 70: 16


In [57]:
anomaly = temps - temps.mean()

# (a) dates more than 3 °C above the mean
print("(a) Days > 3 °C above mean:")
print(dates[anomaly > 3])

# (b) count of days more than 3 °C below the mean
print(f"(b) Days > 3 °C below mean: {(anomaly < -3).sum()}")

(a) Days > 3 °C above mean:
[46272 46274 46276 46277 46279 46281 46282 46284 46287]
(b) Days > 3 °C below mean: 8


In [58]:
first = temps[:30].mean()
last = temps[30:].mean()

print(f"First 30 days: {first:.2f} °C")
print(f"Last 30 days:  {last:.2f} °C")

# Kigali got COOLER: the average dropped by about 4.4 °C from September to October.

First 30 days: 26.21 °C
Last 30 days:  21.78 °C


In [59]:
wettest = sorted(zip(dates, rain), key=lambda p: p[1], reverse=True)[:5]

for d, mm in wettest:
    print(f"{d}: {mm} mm")

46272: 26.2 mm
46279: 25.1 mm
46286: 24.0 mm
46270: 22.9 mm
46293: 22.9 mm


1. Numpy temps.mean() hides the whole mechanism and does it in one line, it hides the loop and the manual sum / count.
2. It was important to learn loops because i understood what mean actually does and if I get a bug, I know where it came from.
3. Kigali got cooler over the period, from 26.21°C in the first 30 days to 21.78°C in the last 30 days,the top 5 days were 22.9 to 26.2mm each. Rainy days were slightly warmer (24.14°C) than dry days (23.95°C), so rain doesn't mean cold here. I found 5 days that were both hot (>=26°C) and rainy, 16 humid days >=70% humidity, and 8 days more than 3°C above the mean and 8 days more than 3°C below, showing the temperature is mostly stable around the mean with a few extreme days.